File: BiPro-Webspace Spiegelung Live/api/bipro_events.php
<?php
/**
 * BiPro API - BiPRO Events
 *
 * Strukturierte Metadaten aus 0-Dokument-Lieferungen
 * (Vertragsdaten-XML, Statusmeldungen, GDV-Ankuendigungen).
 *
 * Endpunkte:
 * - GET    /bipro-events              Paginierte Liste (Filter: event_type, vu_name, is_read)
 * - GET    /bipro-events/summary      Leichtgewichtig fuer Polling (unread_count)
 * - POST   /bipro-events              Neuen Event erstellen
 * - PUT    /bipro-events/read         Bulk-Read-Markierung (IDs)
 * - PUT    /bipro-events/read-all     Alle als gelesen markieren
 * - DELETE /bipro-events/{id}         Einzelnen Event loeschen (Admin)
 * - DELETE /bipro-events/all          Alle Events loeschen (Admin)
 */

require_once __DIR__ . '/lib/db.php';
require_once __DIR__ . '/lib/response.php';
require_once __DIR__ . '/lib/jwt.php';

function handleBiproEventsRequest(string $action, string $method): void {
    $payload = JWT::requireAuth();

    if ($method === 'GET' && (empty($action) || $action === 'list')) {
        listBiproEvents($payload);
        return;
    }

    if ($method === 'GET' && $action === 'summary') {
        getBiproEventsSummary($payload);
        return;
    }

    if ($method === 'POST' && empty($action)) {
        createBiproEvent($payload);
        return;
    }

    if ($method === 'PUT' && $action === 'read') {
        markBiproEventsRead($payload);
        return;
    }

    if ($method === 'PUT' && $action === 'read-all') {
        markAllBiproEventsRead($payload);
        return;
    }

    if ($method === 'DELETE' && $action === 'all') {
        require_once __DIR__ . '/lib/permissions.php';
        requireAdmin();
        deleteAllBiproEvents($payload);
        return;
    }

    if ($method === 'DELETE' && is_numeric($action)) {
        require_once __DIR__ . '/lib/permissions.php';
        requireAdmin();
        deleteBiproEvent($payload, (int)$action);
        return;
    }

    json_error('Methode nicht erlaubt', 405);
}

/**
 * GET /bipro-events
 */
function listBiproEvents(array $user): void {
    $pdo = db();
    $conditions = ['1=1'];
    $params = [];

    if (!empty($_GET['event_type'])) {
        $conditions[] = 'event_type = ?';
        $params[] = $_GET['event_type'];
    }

    if (!empty($_GET['vu_name'])) {
        $conditions[] = 'vu_name LIKE ?';
        $params[] = '%' . $_GET['vu_name'] . '%';
    }

    if (isset($_GET['is_read'])) {
        $conditions[] = 'is_read = ?';
        $params[] = (int)$_GET['is_read'];
    }

    $page = max(1, (int)($_GET['page'] ?? 1));
    $perPage = min(500, max(1, (int)($_GET['per_page'] ?? 50)));
    $offset = ($page - 1) * $perPage;

    $where = implode(' AND ', $conditions);

    $countStmt = $pdo->prepare("SELECT COUNT(*) FROM bipro_events WHERE $where");
    $countStmt->execute($params);
    $total = (int)$countStmt->fetchColumn();

    $stmt = $pdo->prepare("
        SELECT * FROM bipro_events
        WHERE $where
        ORDER BY created_at DESC
        LIMIT $perPage OFFSET $offset
    ");
    $stmt->execute($params);
    $events = $stmt->fetchAll(PDO::FETCH_ASSOC);

    json_response([
        'data' => $events,
        'pagination' => [
            'page' => $page,
            'per_page' => $perPage,
            'total' => $total,
            'total_pages' => (int)ceil($total / $perPage),
        ],
    ]);
}

/**
 * GET /bipro-events/summary
 */
function getBiproEventsSummary(array $user): void {
    $pdo = db();

    $stmt = $pdo->query("SELECT COUNT(*) FROM bipro_events WHERE is_read = 0");
    $unreadCount = (int)$stmt->fetchColumn();

    $stmt = $pdo->query("
        SELECT * FROM bipro_events
        ORDER BY created_at DESC LIMIT 1
    ");
    $latest = $stmt->fetch(PDO::FETCH_ASSOC) ?: null;

    json_response([
        'unread_count' => $unreadCount,
        'latest_event' => $latest,
    ]);
}

/**
 * POST /bipro-events
 */
function createBiproEvent(array $user): void {
    $data = get_json_body();

    if (empty($data['shipment_id'])) {
        json_error('shipment_id erforderlich', 400);
        return;
    }
    if (empty($data['event_type'])) {
        json_error('event_type erforderlich', 400);
        return;
    }

    $allowed = ['gdv_announced', 'contract_xml', 'status_message', 'document_xml'];
    if (!in_array($data['event_type'], $allowed)) {
        json_error('Ungueltiger event_type', 400);
        return;
    }

    $pdo = db();

    $stmt = $pdo->prepare("
        INSERT INTO bipro_events
            (shipment_id, vu_name, vu_bafin_nr, bipro_category, category_name,
             event_type, vsnr, vn_name, vn_address, sparte, vermittler_nr,
             freitext, kurzbeschreibung, referenced_filename, shipment_date,
             raw_document_id)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        ON DUPLICATE KEY UPDATE
            vu_name = VALUES(vu_name),
            vu_bafin_nr = VALUES(vu_bafin_nr),
            bipro_category = VALUES(bipro_category),
            category_name = VALUES(category_name),
            event_type = VALUES(event_type),
            vsnr = VALUES(vsnr),
            vn_name = VALUES(vn_name),
            vn_address = VALUES(vn_address),
            sparte = VALUES(sparte),
            vermittler_nr = VALUES(vermittler_nr),
            freitext = VALUES(freitext),
            kurzbeschreibung = VALUES(kurzbeschreibung),
            referenced_filename = VALUES(referenced_filename),
            shipment_date = VALUES(shipment_date),
            raw_document_id = VALUES(raw_document_id)
    ");

    $stmt->execute([
        $data['shipment_id'],
        $data['vu_name'] ?? null,
        $data['vu_bafin_nr'] ?? null,
        $data['bipro_category'] ?? null,
        $data['category_name'] ?? null,
        $data['event_type'],
        $data['vsnr'] ?? null,
        $data['vn_name'] ?? null,
        $data['vn_address'] ?? null,
        $data['sparte'] ?? null,
        $data['vermittler_nr'] ?? null,
        $data['freitext'] ?? null,
        $data['kurzbeschreibung'] ?? null,
        $data['referenced_filename'] ?? null,
        $data['shipment_date'] ?? null,
        $data['raw_document_id'] ?? null,
    ]);

    $eventId = $pdo->lastInsertId();
    if (!$eventId) {
        $stmt2 = $pdo->prepare("SELECT id FROM bipro_events WHERE shipment_id = ?");
        $stmt2->execute([$data['shipment_id']]);
        $eventId = $stmt2->fetchColumn();
    }

    json_response(['success' => true, 'id' => (int)$eventId], 201);
}

/**
 * PUT /bipro-events/read
 */
function markBiproEventsRead(array $user): void {
    $data = get_json_body();
    $ids = $data['ids'] ?? [];

    if (empty($ids) || !is_array($ids)) {
        json_error('ids Array erforderlich', 400);
        return;
    }

    $pdo = db();
    $placeholders = implode(',', array_fill(0, count($ids), '?'));
    $stmt = $pdo->prepare("UPDATE bipro_events SET is_read = 1 WHERE id IN ($placeholders)");
    $stmt->execute(array_map('intval', $ids));

    json_response(['success' => true, 'updated' => $stmt->rowCount()]);
}

/**
 * PUT /bipro-events/read-all
 */
function markAllBiproEventsRead(array $user): void {
    $pdo = db();
    $stmt = $pdo->prepare("UPDATE bipro_events SET is_read = 1 WHERE is_read = 0");
    $stmt->execute();

    json_response(['success' => true, 'updated' => $stmt->rowCount()]);
}

/**
 * DELETE /bipro-events/{id}  (Admin only)
 */
function deleteBiproEvent(array $user, int $eventId): void {
    $pdo = db();

    $stmt = $pdo->prepare("SELECT id FROM bipro_events WHERE id = ?");
    $stmt->execute([$eventId]);
    if (!$stmt->fetch()) {
        json_error('Event nicht gefunden', 404);
        return;
    }

    $stmt = $pdo->prepare("DELETE FROM bipro_events WHERE id = ?");
    $stmt->execute([$eventId]);

    json_response(['success' => true, 'deleted' => 1]);
}

/**
 * DELETE /bipro-events/all  (Admin only)
 */
function deleteAllBiproEvents(array $user): void {
    $pdo = db();
    $stmt = $pdo->query("DELETE FROM bipro_events");
    $deleted = $stmt->rowCount();

    json_response(['success' => true, 'deleted' => $deleted]);
}

File: src/api/bipro_events.py
"""
ACENCIA ATLAS - BiPRO Events API Client

Zugriff auf die bipro_events-Tabelle: strukturierte Metadaten
aus 0-Dokument-Lieferungen (Vertragsdaten-XML, Statusmeldungen,
GDV-Ankuendigungen).
"""

import logging
from dataclasses import dataclass
from typing import Optional, List, Dict

logger = logging.getLogger(__name__)


@dataclass
class BiproEvent:
    """Einzelner BiPRO-Event-Eintrag."""
    id: int
    shipment_id: str
    vu_name: Optional[str]
    vu_bafin_nr: Optional[str]
    bipro_category: Optional[str]
    category_name: Optional[str]
    event_type: str
    vsnr: Optional[str]
    vn_name: Optional[str]
    vn_address: Optional[str]
    sparte: Optional[str]
    vermittler_nr: Optional[str]
    freitext: Optional[str]
    kurzbeschreibung: Optional[str]
    referenced_filename: Optional[str]
    shipment_date: Optional[str]
    raw_document_id: Optional[int]
    is_read: bool
    created_at: str

    @classmethod
    def from_dict(cls, d: Dict) -> 'BiproEvent':
        return cls(
            id=int(d['id']),
            shipment_id=d['shipment_id'],
            vu_name=d.get('vu_name'),
            vu_bafin_nr=d.get('vu_bafin_nr'),
            bipro_category=d.get('bipro_category'),
            category_name=d.get('category_name'),
            event_type=d['event_type'],
            vsnr=d.get('vsnr'),
            vn_name=d.get('vn_name'),
            vn_address=d.get('vn_address'),
            sparte=d.get('sparte'),
            vermittler_nr=d.get('vermittler_nr'),
            freitext=d.get('freitext'),
            kurzbeschreibung=d.get('kurzbeschreibung'),
            referenced_filename=d.get('referenced_filename'),
            shipment_date=d.get('shipment_date'),
            raw_document_id=int(d['raw_document_id']) if d.get('raw_document_id') else None,
            is_read=bool(int(d.get('is_read', 0))),
            created_at=d.get('created_at', ''),
        )


class BiproEventsAPI:
    """API-Client fuer BiPRO-Events."""

    def __init__(self, client):
        self.client = client

    def get_events(self, page: int = 1, per_page: int = 20,
                   event_type: str = None, is_read: int = None) -> Dict:
        params = {'page': page, 'per_page': per_page}
        if event_type:
            params['event_type'] = event_type
        if is_read is not None:
            params['is_read'] = is_read
        try:
            resp = self.client.get('/bipro-events', params=params)
            events = [BiproEvent.from_dict(e) for e in resp.get('data', [])]
            return {'data': events, 'pagination': resp.get('pagination', {})}
        except Exception as e:
            logger.error(f"Fehler beim Laden der BiPRO-Events: {e}")
            return {'data': [], 'pagination': {}}

    def get_summary(self) -> Dict:
        try:
            return self.client.get('/bipro-events/summary')
        except Exception as e:
            logger.debug(f"BiPRO-Events Summary Fehler: {e}")
            return {'unread_count': 0, 'latest_event': None}

    def create_event(self, data: Dict) -> Optional[int]:
        try:
            resp = self.client.post('/bipro-events', json_data=data)
            if resp.get('success'):
                return resp.get('id')
        except Exception as e:
            logger.warning(f"BiPRO-Event erstellen fehlgeschlagen: {e}")
        return None

    def mark_as_read(self, ids: List[int]) -> None:
        if not ids:
            return
        try:
            self.client.put('/bipro-events/read', json_data={'ids': ids})
        except Exception as e:
            logger.warning(f"BiPRO-Events als gelesen markieren fehlgeschlagen: {e}")

    def mark_all_read(self) -> int:
        """Markiert alle Events als gelesen. Gibt Anzahl aktualisierter Events zurueck."""
        try:
            resp = self.client.put('/bipro-events/read-all', json_data={})
            return resp.get('updated', 0)
        except Exception as e:
            logger.warning(f"Alle BiPRO-Events als gelesen markieren fehlgeschlagen: {e}")
            return 0

    def delete_event(self, event_id: int) -> bool:
        """Loescht einen einzelnen Event (Admin)."""
        try:
            resp = self.client.delete(f'/bipro-events/{event_id}')
            return resp.get('success', False)
        except Exception as e:
            logger.warning(f"BiPRO-Event {event_id} loeschen fehlgeschlagen: {e}")
            return False

    def delete_all_events(self) -> int:
        """Loescht alle Events (Admin). Gibt Anzahl geloeschter Events zurueck."""
        try:
            resp = self.client.delete('/bipro-events/all')
            return resp.get('deleted', 0)
        except Exception as e:
            logger.warning(f"Alle BiPRO-Events loeschen fehlgeschlagen: {e}")
            return 0
